File: Tutorials/Scala/4 Routing/producer/src/main/scala/ru/example/rmq/routing/Producer.scala
package ru.example.rmq.routing

import com.rabbitmq.client._
import java.nio.charset.StandardCharsets
import scala.util.Using

object Producer {
  def main(args: Array[String]): Unit = {
    val EXCHANGE_NAME = "direct_logs"
    val virtualHost   = "/"

    if (args.length < 3) {
      System.out.println("Usage: Consumer hostname username password [severity] [message...]")
      System.exit(-1)
    }

    val hostname = args(0)
    val username = args(1)
    val password = args(2)

    val severity = if (args.length > 3) args(3) else "info"
    val message  = if (args.length > 4) args.slice(4, args.length).mkString(" ") else "Hello World!"

    val factory = new ConnectionFactory
    factory.setHost(hostname)
    factory.setUsername(username)
    factory.setPassword(password)
    factory.setVirtualHost(virtualHost)

    Using.Manager { use =>
      try {
        val connection = use(factory.newConnection)
        val channel    = use(connection.createChannel)

        channel.exchangeDeclare(EXCHANGE_NAME, BuiltinExchangeType.DIRECT)
        channel.basicPublish(
          EXCHANGE_NAME,
          severity,
          null,
          message.getBytes(StandardCharsets.UTF_8)
        )
        println(s"[x] Sent '$severity': '$message'")
      } catch {
        case e: Exception =>
          println(e.getLocalizedMessage)
          sys.exit(-1)
      }
    }
    sys.exit(0)
  }
}

File: Tutorials/Java/4 Routing/Producer/src/main/java/Producer.java
import com.rabbitmq.client.BuiltinExchangeType;
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;

public class Producer {
    public static void main(String[] argv) throws Exception {
        String hostname;
        String username;
        String password;
        String virtualHost = "/";
        String EXCHANGE_NAME = "direct_logs";
        String severity;
        String message;

        if (argv.length < 3) {
            System.out.println("Usage: Producer hostname username password [severity] [message...]");
            System.exit(-1);
        }

        hostname = argv[0];
        username = argv[1];
        password = argv[2];

        if (argv.length > 3) severity = argv[3];
        else severity = "info";

        if (argv.length > 4) message = String.join(" ", Arrays.copyOfRange(argv, 4, argv.length));
        else message = "Hello World!";

        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost(hostname);
        factory.setUsername(username);
        factory.setPassword(password);
        factory.setVirtualHost(virtualHost);

        try (Connection connection = factory.newConnection();
             Channel channel = connection.createChannel()) {
            channel.exchangeDeclare(EXCHANGE_NAME, BuiltinExchangeType.DIRECT);
            channel.basicPublish(EXCHANGE_NAME, severity, null, message.getBytes(StandardCharsets.UTF_8));
            System.out.println(" [x] Sent '" + severity + "':'" + message + "'");
        }
    }
}